docs/geps/gep-04.rst

```diff
@@ -0,0 +1,356 @@
+GETTSIM's scope. There has to be some discretion in order to allow for the interface of
+functions to change over time, new functions to appear, or old ones to disappear.
+
+Some examples include:
+
+1. ``arbeitsl_hilfe`` being replaced by ``arbeitsl_geld_2``
+1. ``kinderbonus`` being active only in a few years
+1. The introduction of ``kinderzuschlag``
+1. Capital income entering ``sum_brutto_eink`` or not.
+
+The goal is that the graph for any particular point in time is minimal in the sense that
+``arbeitsl_geld_2`` does not appear before it was conceived, it is apparent from the
+interface of ``sum_brutto_eink`` whether it includes capital income or not, etc..
+
+In the yaml-files corresponding to a particular tax / transfer, functions not present in
+all years will need to be listed with along with the dates for when they are active. See
+:gep-3-keys-referring-to-functions: for the precise syntax. That mechanism should be
+used for:
+
+1. Functions that are newly introduced.
+2. Functions that cease to be relevant.
+3. Functions whose interface changes over time.
+4. Functions whose body changes so much that
+
+   - it is useful to signal that things have changed and/or
+
+   - it would be awkward to program the different behaviors in one block with case
+     distinctions.
+
+Needless to say, the different reasons may appear at different points in time for the
+same function.
+
+
+Additional functionalities
+--------------------------
+
+We implemented a small set of additional features that simplify the specification of
+certain types of functions of the taxes and transfers system.
+
+
+.. _gep-4-aggregation-functions:
+
+Group summation and other aggregation functions
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Many taxes or transfers require group-level variables. <GEP-2 describes
+`gep-2-aggregation-functions`> how reductions are handled in terms of the underlying
+data. This section describes how to specify them.
+
+In order to inject aggregation functions into the graph, scripts with functions of the
+taxes and transfer system should define a dictionary ``aggregation_[script_name]`` at
+the module level. This dictionary must specify the aggregated columns as keys and a
+dictionary with keys ``source_col`` and ``aggr`` as values. If ``aggr`` is ``count``,
+``source_col`` is not needed.
+
+
+For example, in ``demographic_vars.py``, we could have:
+
+.. code::
+
+    aggregation_demographic_vars = {
+        "anz_erwachsene_tu": {"source_col": "erwachsen", "aggr": "sum"},
+        "haushaltsgröße_hh": {"aggr": "count"},
+    }
+
+The group identifier (``tu_id``, ``hh_id``) will be automatically included as an
+argument; for ``count`` no other variable is necessary.
+
+The output type will be the same as the input type. Exceptions:
+
+- Input type ``bool`` and aggregation ``sum`` leads to output type ``integer``.
+- Input type ``integer`` and agggregation :math:`\in \{` ``any``, ``all`` :math:`\}`
```

Comment:
Maybe add here:

- Input type ``integer`` and aggregation ``mean`` leads to output type ``float``
- Aggregation ``count``  always leads to output type ``integer``

Also: I used ``int`` instead of ``integer`` for type annotations so far.

Reply:
Will do. Do we have examples of taking means of integers?

Reply:
> Do we have examples of taking means of integers?

No, we don't use ``mean`` aggregation at all at the moment.

Reply:
Cool. Then we should only allow it for floats in the first place. Will always be able to change that down the road should it come up.